### PR TITLE
Upgrade suite 4.x to 5.x for cephfs

### PR DESCRIPTION
**Pull request by AmarnatReddy**
Upgrade suite 4.x to 5.x for cephfs
This covers
1.  creation of subvolumegroup,subvolume,snapshot,clone 
2. writing data to subvolume
3. Creation of nfs server and mounting 
4. Writing data to nfs server
5. Validating the file permissions/stats before and after the upgrade

Logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-A89QZR/ 

Latest Log : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-DWN3E8/

Signed-off-by: Amarnath K <amk@amk.remote.csb>

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by hkadam134** [comment]
@AmarnatReddy 
Can we specify the exact version of upgrades in the file names?

File: tests/cephfs/cephfs_upgrade/cephfs_io.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        if not clients:
+            log.info(
+                f"This test requires minimum 1 client nodes.This has only {len(clients)} clients"
+            )
+            return 1
+        client1 = clients[0]
+        stats = {"total_iterations": 0, "smallfile": 0, "dd": 0}
+        out, rc = client1.exec_command(
+            sudo=True, cmd="mount -t fuse.ceph-fuse | awk {'print $3'}", check_ec=False
+        )
```
Comment (hkadam134) [inline tests/cephfs/cephfs_upgrade/cephfs_io.py:29]: Can we also have IO's running on kernel mounts?

File: tests/cephfs/cephfs_upgrade/upgrade_pre_req.py [hunk 2]
```diff
@@ -116,73 +125,145 @@ def run(ceph_cluster, **kw):
...
-                sudo=True,
-                cmd=f"ceph nfs export create cephfs {fs_name} {nfs_name} "
-                f"{nfs_export_name} path={export_path}",
+            if "nautilus" not in ceph_version["version"]:
+                fs_util.fuse_mount(
+                    [clients[0]],
+                    fuse_mounting_dir_1,
+                    extra_params=f"-r {subvol_path.strip()} --client_fs {sv['vol_name']}",
+                )
+                mount_points["fuse_mounts"].append(fuse_mounting_dir_1)
```
Comment (yogesh-mane) [inline tests/cephfs/cephfs_upgrade/upgrade_pre_req.py:134]: Could we add code for mounting ceph-fuse for 4.x here?

File: tests/cephfs/cephfs_upgrade/upgrade_pre_req.py [hunk 3]
```diff
@@ -116,73 +125,145 @@ def run(ceph_cluster, **kw):
...
+                sudo=True,
+                cmd=f"python3 /home/cephuser/smallfile/smallfile_cli.py --operation read --threads 10 --file-size 4 "
+                f"--files 1000 --files-per-dir 10 --dirs-per-dir 2 --top "
+                f"{nfs_mounting_dir}{dir_name}",
+                long_running=True,
+            )
+
+        else:
+
+            clients = ceph_cluster.get_ceph_objects("client")
```
Comment (yogesh-mane) [inline tests/cephfs/cephfs_upgrade/upgrade_pre_req.py:217]: I am not seeing this nfs block getting executed in logs. Could you check that once?

File: suites/pacific/cephfs/tier-2_cephfs_upgrade.yaml [hunk 4]
```diff
@@ -0,0 +1,106 @@
+---
+tests:
+  -
+    test:
+      abort-on-fail: true
+      module: install_prereq.py
+      name: "install ceph pre-requisites"
```
Comment (yogesh-mane) [inline suites/pacific/cephfs/tier-2_cephfs_upgrade.yaml:7]: Could we have test description here?

File: suites/pacific/cephfs/tier-2_cephfs_upgrade.yaml [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        use_cdn: true
+      desc: "deploy ceph containerized 4.x cdn setup using ceph-ansible"
+      destroy-cluster: false
+      module: test_ansible.py
+      name: "ceph ansible install containerized rhcs 4.x from cdn"
+      polarion-id: CEPH-83573588
+  -
+    test:
+      abort-on-fail: false
+      desc: ~
```
Comment (yogesh-mane) [inline suites/pacific/cephfs/tier-2_cephfs_upgrade.yaml:52]: Could we have description here & for next test as well?

File: tests/cephfs/cephfs_upgrade/upgrade_pre_req.py [hunk 6]
```diff
@@ -29,6 +29,7 @@ def run(ceph_cluster, **kw):
         fs_util = FsUtils(ceph_cluster)
         config = kw.get("config")
         build = config.get("build", config.get("rhbuild"))
+        build = "4.3"
```
Comment (yogesh-mane) [inline tests/cephfs/cephfs_upgrade/upgrade_pre_req.py:32]: Could we remove this?

File: tests/cephfs/cephfs_upgrade/cephfs_io.py (hunk 1 above)
Comment (AmarnatReddy) [reply tests/cephfs/cephfs_upgrade/cephfs_io.py:30]: Done

File: tests/cephfs/cephfs_upgrade/upgrade_pre_req.py [hunk 7]
```diff
@@ -28,7 +28,7 @@ def run(ceph_cluster, **kw):
     try:
         fs_util = FsUtils(ceph_cluster)
         config = kw.get("config")
-        build = config.get("build", config.get("rhbuild"))
+        build = config.get("build", "4.3")
```
Comment (hkadam134) [inline tests/cephfs/cephfs_upgrade/upgrade_pre_req.py:31]: Is this okay to hardcode a version ?

**Review by hkadam134 — APPROVED** [review]
Looks Good

File: tests/cephfs/cephfs_upgrade/upgrade_pre_req.py (hunk 2 above)
Comment (yogesh-mane) [reply tests/cephfs/cephfs_upgrade/upgrade_pre_req.py:133]: Just one thing, I am not seeing ceph-fuse mounts in latest logs.

File: tests/cephfs/cephfs_upgrade/upgrade_pre_req.py (hunk 3 above)
Comment (yogesh-mane) [reply tests/cephfs/cephfs_upgrade/upgrade_pre_req.py:223]: Just one thing, I am not seeing ceph-fuse mounts in latest logs.

**Review by yogesh-mane — COMMENTED** [review]
Just one thing, I am not seeing ceph-fuse mounts in latest logs for upgrade_pre_req script.

**Comment by AmarnatReddy** [comment]
> Just one thing, I am not seeing ceph-fuse mounts in latest logs for upgrade_pre_req script.

Hi Yogesh,
Volumes are created on cephfs and cephfs-ec.
when i split the list of volumes, which I tried for fuse mounts were falling in cephfs-ec.
In 4. x we don't have support for --client-fs argument.

**Review by yogesh-mane — APPROVED** [review]
Looks finr

**Pull request merged**